server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java

```diff
@@ -527,22 +527,31 @@ private void startElection() {
     }
 
     private void broadcastStartJoinRequest(StartJoinRequest startJoinRequest, List<DiscoveryNode> discoveredNodes) {
-        electionStrategy.onNewElection(
-            startJoinRequest.getSourceNode(),
-            startJoinRequest.getTerm(),
-            getLastAcceptedState(),
-            new ActionListener<>() {
-                @Override
-                public void onResponse(Void ignored) {
-                    discoveredNodes.forEach(node -> joinHelper.sendStartJoinRequest(startJoinRequest, node));
+        electionStrategy.onNewElection(startJoinRequest.getSourceNode(), startJoinRequest.getTerm(), new ActionListener<>() {
+            @Override
+            public void onResponse(ClusterState latestClusterState) {
+                if (latestClusterState != null) {
+                    if (latestClusterState.term() > getCurrentTerm()) {
+                        updateCurrentTerm(latestClusterState.term());
+                    }
+                    coordinationState.get().setLatestAcceptedState(latestClusterState);
+                    preVoteCollector.update(getPreVoteResponse(), null);
                 }
+                discoveredNodes.forEach(node -> joinHelper.sendStartJoinRequest(startJoinRequest, node));
+            }
 
-                @Override
-                public void onFailure(Exception e) {
-                    logger.debug(Strings.format("election attempt failed, dropping [%s]", startJoinRequest), e);
-                }
+            @Override
+            public void onFailure(Exception e) {
+                logger.debug(Strings.format("election attempt failed, dropping [%s]", startJoinRequest), e);
             }
-        );
+        });
+    }
+
+    private void updateCurrentTerm(long term) {
```

Comment:
I'm not a fan of this, another option is to return here the previous master + the latest cluster state and call `ensureTermAtLeast` but that might be tricky to get right? i.e. for now we're not storing the nodes in the store

Reply:
I wonder if it would be better to do this later on. I mean if we don't bump the term here, we send out the start-join requests which triggers some join requests which themselves bump the term.

In fact now that I write that, this seems to be the wrong place to read the fresher cluster state. In the `abdicateTo()` case we send out a start-join request on behalf of the node to which we're abdicating, and the target node receives the join requests. So we should be reading the new state when handling those join requests, i.e. somewhere like `CandidateJoinAccumulator#close`